### ISSUE TITLE
Creating discriminators can have unintended side effects, including modifying unrelated Model's schemas

### ISSUE THREAD

**Issue by samgladstone** [Automattic/mongoose#8546]
**Do you want to request a *feature* or report a *bug*?** Bug

**What is the current behavior?**
Creating a model discriminator can have unintended side effects, including modifying the schemas of completely unrelated models

**If the current behavior is a bug, please provide the steps to reproduce.**

I wrote these tests, that recreate the issue. I believe they should be added to: [/test/model.discriminator.test.js line 1326](https://github.com/Automattic/mongoose/blob/master/test/model.discriminator.test.js#L1326), which would result in description "_model / discriminator() / does not have unintended side effects / when the base schema has an \_id that is not auto generated_"

```js
describe('does not have unintended side effects', function () {
    // Delete every model
    afterEach(function () { mongoose.deleteModel(/.+/); });

    function throwErrorOnClone() { throw new Error('clone() was called on the unrelated schema'); };

    it('when the base schema has an _id that is not auto generated', function() {
        const unrelatedSchema = new mongoose.Schema({});
        unrelatedSchema.clone = throwErrorOnClone;
        mongoose.model('UnrelatedModel', unrelatedSchema);

        const model = mongoose.model('Model', new mongoose.Schema({ _id: mongoose.Types.ObjectId }, { _id: false }));
        model.discriminator('Discrimintaor', new mongoose.Schema({}).clone());
    });

    it('when using a new mongoose instance', function() {
        const mongooseInstance = new mongoose.Mongoose();

        const unrelatedSchema = new mongooseInstance.Schema({});
        unrelatedSchema.clone = throwErrorOnClone;
        mongooseInstance.model('UnrelatedModel', unrelatedSchema);

        const model = mongooseInstance.model('Model', new mongooseInstance.Schema({}));
        model.discriminator('Discriminator', new mongooseInstance.Schema({}))
    });

    it('when the base schema has been cloned', function() {
        const unrelatedSchema = new mongoose.Schema({});
        unrelatedSchema.clone = throwErrorOnClone;
        mongoose.model('UnrelatedModel', unrelatedSchema);

        const model = mongoose.model('Model', new mongoose.Schema({}).clone());
        model.discriminator('Discriminator', new mongoose.Schema({}))
    });
});
```

**What is the expected behavior?**
These side effects should not occur.

**Versions**

Node: 12.14.1
Mongoose: 5.8.9
Mongo: 4.2.1

**Comment by samgladstone** [comment]
I've spent some time investigating the issue and have the following thoughts/observations. 

### When the base schema has an _id that is not auto generated

The first test that fails is caused by what I believe is unnecessary code, I created a pull request for this fix in #8543 and will now add this test case.

### Others

Following through the code:

The base schema is merged into the discriminating schema using `utils.merge` in [/helpers/model/discriminator.js](https://github.com/Automattic/mongoose/blob/master/lib/helpers/model/discriminator.js#L92)

```js
    utils.merge(schema, baseSchema, {
      omit: { discriminators: true },
      omitNested: conflictingPaths.reduce((cur, path) => {
        cur['tree.' + path] = true;
        return cur;
      }, {})
    });
```

The merge in [utils.js](https://github.com/Automattic/mongoose/blob/master/lib/utils.js#L400) uses `Object.keys(from)` to get a list of properties to merge. In the cases that fail `baseSchema.base` is being included in the merge. This, in turn, means that all existing model schemas are merged with themselves.

```js
exports.merge = function merge(to, from, options, path) {
  options = options || {};

  const keys = Object.keys(from);
```

### This doesn't happen in every case
`Schema.base` is not included when using _standard_ schemas as it is defined on the prototype (in [index.js line 98](https://github.com/Automattic/mongoose/blob/master/lib/index.js#L98) and so is not included in `Object.keys(from)`

```js
this.Schema.prototype.base = this;
```

### When using a new mongoose instance
`Schema.base` is set against the instance of Schema when a new mongoose instance is created so is included in the merge:

```js
// If a user creates their own Mongoose instance, give them a separate copy
  // of the `Schema` constructor so they get separate custom types. (gh-6933)
  if (!options || !options[defaultMongooseSymbol]) {
    const _this = this;
    this.Schema = function() {
      this.base = _this;
      return Schema.apply(this, arguments);
    };
    this.Schema.prototype = Object.create(Schema.prototype);

    Object.assign(this.Schema, Schema);
    this.Schema.base = this;
    this.Schema.Types = Object.assign({}, Schema.Types);
  } else {
    // Hack to work around babel's strange behavior with
    // `import mongoose, { Schema } from 'mongoose'`. Because `Schema` is not
    // an own property of a Mongoose global, Schema will be undefined. See gh-5648
    for (const key of ['Schema', 'model']) {
      this[key] = Mongoose.prototype[key];
    }
  }
  this.Schema.prototype.base = this;
```

#### Causes

I found that this issue was introduced across related issues:

#6933 - commit 476c1b9031f01a42a9e355b34f88f55e38fb7dea
#7158 - commit c7f216cb4349ea157ac3279cb6fd32c88892cfa8
#7103 - commit 0ef6d1627b45a68c6cf58b978b921b27e0de9e0a

### When the base schema has been cloned

When cloning a schema in [schema.js line 296](https://github.com/Automattic/mongoose/blob/master/lib/schema.js#L296), the base is set against the cloned instance of the schema.

```js
Schema.prototype.clone = function() {
  const s = new Schema({}, this._userProvidedOptions);
  s.base = this.base;
```

This was introduced in the fix for #7377 - commit cdf89dd44ada368ba6dd8a2f912dafd5a89a8982

### Possible Solutions

#### Exclude base from the merge

This seems very straightforward and also matches the _standard_ behaviour (change in [/helpers/model/discriminator.js line 92](https://github.com/Automattic/mongoose/blob/master/lib/helpers/model/discriminator.js#L92))

```js
 utils.merge(schema, baseSchema, {
      omit: {
        discriminators: true,
        base: true,  // ADDED
      },
      omitNested: conflictingPaths.reduce((cur, path) => {
        cur['tree.' + path] = true;
        return cur;
      }, {})
    });
```
My concern is that there may be a situation where you want the bases to be matching. I am left wondering if it would be helpful to have a check to ensure that schemas that are being merged to create a discriminator should have the same base?

#### Use base to create schema when cloning

Another option that fixes only the _when the base schema has been cloned_  case is to use the Schema's base when cloning:

```js
Schema.prototype.clone = function() {
  const s = new this.base.Schema({}, this._userProvidedOptions);
  s.obj = this.obj;
  s.options = utils.clone(this.options);
```

This passes the tests introduced in 43851c791a7cdf3b1c49a4b04e39c8eef4718cff and seems like a less brittle approach.

#### Simplify adding base to schema instances

I believe that removing the following has no bad side effects and solves the _when using a new mongoose instance_ issue (removed lines are commented with `// REMOVED`:

```js
// If a user creates their own Mongoose instance, give them a separate copy
  // of the `Schema` constructor so they get separate custom types. (gh-6933)
  if (!options || !options[defaultMongooseSymbol]) {
    //const _this = this;        // REMOVED
    this.Schema = function() {
      //this.base = _this;       // REMOVED
      return Schema.apply(this, arguments);
    };
    this.Schema.prototype = Object.create(Schema.prototype);

    Object.assign(this.Schema, Schema);
    //this.Schema.base = this;            // REMOVED
    this.Schema.Types = Object.assign({}, Schema.Types);
  } else {
    // Hack to work around babel's strange behavior with
    // `import mongoose, { Schema } from 'mongoose'`. Because `Schema` is not
    // an own property of a Mongoose global, Schema will be undefined. See gh-5648
    for (const key of ['Schema', 'model']) {
      this[key] = Mongoose.prototype[key];
    }
  }
  this.Schema.prototype.base = this;
```
As `this.Schema` gets assigned a new function at the beginning of that snippet, assigning `this.Schema.prototype.base = this;` is equivalent to the removed lines of code. At least, as far as I can make out, unless there is something that would modify it later. This passes all the tests from the causing issues mentioned above (#6933, #7158, #7103)

### Why the side effects are an issue

Setting aside, whether or not `Schema.base` should be merged, I would expect that merging a schema with itself would have no effect, however, the schema can be modified in certain cases. I tracked the issue down to following code in [util.merge()]():

```js
if (from[key] != null) {
    if (from[key].instanceOfSchema) {
        if (to[key].instanceOfSchema) {
            to[key].add(from[key].clone()); // HERE
        } else {
            to[key] = from[key].clone();
        }
```

At first, I realised that `SchemaType.clone()` is not accounting for any changes made since creation. I have added a fix for that in #8542. 

However, this did not resolve my issue as `Schema.add()` also does not account for changes to the schema paths. Example of testing this (which I think should be added to [/test/schema.test.js line 1303](https://github.com/Automattic/mongoose/blob/master/test/schema.test.js#L1303)):

```js
it('accounts for changes to a schema path', function() {
    const defaultFunction = () => 'abc';

    const schema = new mongoose.Schema({ value: String });
    let path = schema.path('value');
    
    // Changes that can happen
    path.default(defaultFunction);
    path.get(v => v.trim());
    path.index(true);
    path.ref('User');
    path.required(true);
    path.select(false);
    path.set(v => v.trim());
    path.sparse(true);
    path.text(true);
    path.unique(true);
    path.validate(v => v.length > 3);
    path.cast(v => v.length > 3 ? v : v.trim());

    // This should have no side-effect
    schema.add(schema);

    assert.deepStrictEqual(schema.path('value'), path);

    // immutable cannot be tested with deep equals
    path = schema.path('value');
    path.immutable(true);

    schema.add(schema);
    let newPath = schema.path('value');
    
    assert.equal(path.$immutable, newPath.$immutable);
    assert.equal(path.setters.length, newPath.setters.length);
});
```
Following the code to see why this is happening:

`Schema.add()` calls `merge()` in [/helpers/schema/merge](https://github.com/Automattic/mongoose/blob/master/lib/helpers/schema/merge.js). 

```js
Schema.prototype.add = function add(obj, prefix) {
  if (obj instanceof Schema) {
    merge(this, obj);
    return this;
  }
```

In turn, `merge()` uses the schema's tree to do the merge:

```js
module.exports = function merge(s1, s2) {
  s1.add(s2.tree || {});
```

The schema's tree is not updated when modifying the schema paths. I found this to be a harder problem to solve and have not really attempted to tackle it. I did wonder if merging `s2.paths` with `s1.paths` after `s1.add(s2.tree || {});` would resolve the issue, but I haven't tried that

Perhaps this is a separate issue? 

Finally, I'll happily create another pull request with the above changes if that would be helpful to you?

**Comment by vkarpov15** [comment]
In the end I decided to just `remove('_id')` and get rid of the unnecessary `merge()`. I intended to avoid mutating the passed in schema to `Model.discriminator()`, but that doesn't look feasible right now with how we implement discriminators.

I like your suggestion to clone schemas when passing to `Model.discriminator()` to avoid mutating the schema, but you're right, that causes a lot of test failures that make me think that cloning on every `discriminator()` call is dangerous. We'll consider that for a future release.
